cibuildwheel/macos.py

```diff
@@ -395,202 +334,202 @@ def build(options: Options) -> None:
-                                "don't know how to emulate {testing_arch} on {machine_arch}"
-                            )
+                        # set up a virtual environment to install and test from, to make sure
+                        # there are no dependencies that were pulled in at build time.
+                        call("pip", "install", "virtualenv", *dependency_constraint_flags, env=env)
+                        venv_dir = Path(tempfile.mkdtemp())
+
+                        arch_prefix = []
+                        if testing_arch != machine_arch:
+                            if machine_arch == "arm64" and testing_arch == "x86_64":
+                                # rosetta2 will provide the emulation with just the arch prefix.
+                                arch_prefix = ["arch", "-x86_64"]
+                            else:
+                                raise RuntimeError(
+                                    "don't know how to emulate {testing_arch} on {machine_arch}"
+                                )
 
-                    # define a custom 'call' function that adds the arch prefix each time
-                    def call_with_arch(args: Sequence[PathOrStr], **kwargs: Any) -> None:
-                        if isinstance(args, str):
-                            args = " ".join(arch_prefix) + " " + args
-                        else:
-                            args = [*arch_prefix, *args]
-                        call(args, **kwargs)
-
-                    # Use --no-download to ensure determinism by using seed libraries
-                    # built into virtualenv
-                    call_with_arch(
-                        ["python", "-m", "virtualenv", "--no-download", venv_dir], env=env
-                    )
+                        # define a custom 'call' function that adds the arch prefix each time
+                        def call_with_arch(*args: PathOrStr, **kwargs: Any) -> None:
+                            call(*arch_prefix, *args, **kwargs)
 
-                    virtualenv_env = env.copy()
-                    virtualenv_env["PATH"] = os.pathsep.join(
-                        [
-                            str(venv_dir / "bin"),
-                            virtualenv_env["PATH"],
-                        ]
-                    )
+                        def shell_with_arch(command: str, **kwargs: Any) -> None:
+                            command = " ".join(arch_prefix) + " " + command
+                            shell(command, **kwargs)
 
-                    # check that we are using the Python from the virtual environment
-                    call_with_arch(["which", "python"], env=virtualenv_env)
+                        # Use --no-download to ensure determinism by using seed libraries
+                        # built into virtualenv
+                        call_with_arch(
+                            "python", "-m", "virtualenv", "--no-download", venv_dir, env=env
```

Comment:
I think there is still some refactor to be done.
I'm not sure if it's worth pursuing now or later.
Another thing I think would be good to have is some kind of temp_path factory that would things to get  clean at the end of cibuildwheel run (with proper reporting of files not deleted, at least when run locally)

Reply:
re-thinking about this.
for now, it's not the same virtualenv that's called between build (virtualenv dependency of cibuildwheel running with the same interpreter as cibuildwheel) and tests (pinned virtualenv installed in the build venv running with the build venv interpreter)

Reply:
> Another thing I think would be good to have is some kind of temp_path factory that would things to get clean at the end of cibuildwheel run (with proper reporting of files not deleted, at least when run locally).

Agreed. It could attempt deletion, log failures but still continue.

Reply:
> for now, it's not the same virtualenv that's called between build (virtualenv dependency of cibuildwheel running with the same interpreter as cibuildwheel) and tests (pinned virtualenv installed in the build venv running with the build venv interpreter)

Ahhhh, this is a little subtle. I think I see what you mean. The build venv is created using virtualenv running on the 'host' python, with the virtualenv version from `setup.cfg`. The test venv is created using the installed CPython, with the virtualenv version from `constraints.txt`. 

Hmm.

Reply:
> for now, it's not the same virtualenv that's called between build (virtualenv dependency of cibuildwheel running with the same interpreter as cibuildwheel) and tests (pinned virtualenv installed in the build venv running with the build venv interpreter)

Ahhhh, this is a little subtle. I think I see what you mean. The build venv is created using virtualenv running on the 'host' python, with the virtualenv version from `setup.cfg`. The test venv is created using the installed CPython, with the virtualenv version from `constraints.txt`. 

Hmm. Actually I'm wondering if this is right. This might introduce some indeterminism into the build process, because the virtualenv version that creates the build venv would be whatever is available at install time. This is liable to change as time goes on, even if people pin their cibuildwheel version. I can imagine this creating issues down the line. Would it be possible to somehow pin this version of virtualenv?

Reply:
> Would it be possible to somehow pin this version of virtualenv?

Yes it could, but this also raises another question, shall all (direct and transitive) dependencies of cibuildwheel be pinned ? As it's a full standalone application, it might make sense for full reproducibility but I'll let @henryiii enlighten us about pinning everything given its recent posts about this.

Reply:
>> Another thing I think would be good to have is some kind of temp_path factory that would things to get clean at the end of cibuildwheel run (with proper reporting of files not deleted, at least when run locally).

> Agreed. It could attempt deletion, log failures but still continue.

I have something for tmp that I'll get into this. It'll probably simplify reading the diff, especially if/when #978 gets merged.

Reply:
> shall all (direct and transitive) dependencies of cibuildwheel be pinned?

Personally, I think I'd be okay with this. Users can always use pipx if they have issues. But we could just pin virtualenv too. The rest of the dependencies don't have a big impact on the build environment I think.

Reply:
We can provide a lock file of some sort if people really want full pinning, but I would not release a fully pinned package on PyPI. There are just way too many things that can go wrong, and unless we keep up constant releases, things are far more likely to break than work. End users can use a lock file if they want full reproducibility, just like they should do with any other package - if they want this, they can get it using tools like pip-tools, PDM, or poetry with cibuildwheel. Since there's not a standard format for a lock file yet, I'd probably wait to add our own suggested lock file.

Reply:
I pinned virtualenv only for now. If users aren't installing cibuildwheel locally in its own venv, then it's likely not to be respected after a few days/weeks.
I'll check if https://github.com/pypa/get-virtualenv is an option here. It might impact speed (that I'll check easily) and things might just break for now (https://github.com/pypa/virtualenv/issues/2133) so workaround needed or fix to virtualenv.

Reply:
https://github.com/pypa/get-virtualenv is definitely an option here. It doesn't impact speed and there a simple workaround to the issue I mentioned: https://github.com/pypa/virtualenv/issues/2133#issuecomment-1003710125
I need to figure out a few details about how to manage the version update but I think this might be the way to go.

Reply:
virtualenv version management is now handled with its own toml file (along with necessary tools to update)

Reply:
Great!